lib/plugins/cleaner.rb

```diff
@@ -0,0 +1,51 @@
+require 'rubygems/package'
+require_relative '../moonshot/creds_helper'
+
+module Moonshot
+  module Plugins
+    # Moonshot plugin class for removing artifacts when a stage gets deleted.
+    class Cleaner
+      include Moonshot::CredsHelper
+
+      def initialize(bucket_name)
+        @bucket_name = bucket_name
+      end
+
+      def post_delete(resources)
+        stack_name = resources.stack.name
+        resources.ilog.start 'Starting to delete artifacts for stack: ' \
+          "#{stack_name}" do |s|
+          arts = retrieve_artifact_names(stack_name)
+          if arts.empty?
+            s.success 'No artifacts found, nothing deleted.'
+          else
+            delete_artifacts arts
+            s.success "Deleted artifacts for stack: '#{stack_name}' " \
+                      "in bucket: '#{@bucket_name}'"
+          end
+        end
+      end
+
+      private
+
+      def retrieve_artifact_names(name)
+        objs = s3_client.list_objects(bucket: @bucket_name, prefix: name)
+        valid_keys = []
+        objs.contents.each do |o|
+          valid_keys << o.key if o.key =~ /^#{name}[\.-]/
```

Comment:
**[Theoretical level]**

I know it's a much harder way to manage and update the codebase, but what if we can store artifacts with directory stucture?

**Now:**
```
/dev-chris-version1.tar.gz
/dev-chris-version2.tar.gz
/dev-chris-test1-version1.tar.gz
/dev-chris-test1-version2.tar.gz
```

**After:**
```
/dev-chris/dev-chris-version1.tar.gz
/dev-chris/dev-chris-version2.tar.gz
/dev-chris-test1/dev-chris-test-version1.tar.gz
/dev-chris-test1/dev-chris-test-version2.tar.gz
```

With this structure we can easily find `dev-chris` and delete artifacts under that directory without conflict on `dev-chris-test1`